core/lib/zksync_core/src/state_keeper/io/seal_logic/l2_block_seal_subtasks.rs

```diff
@@ -0,0 +1,542 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use futures::FutureExt;
+use zksync_dal::{Core, CoreDal};
+use zksync_db_connection::{connection::Connection, connection_pool::ConnectionPool};
+use zksync_types::{event::extract_added_tokens, L2BlockNumber};
+
+use crate::state_keeper::{
+    metrics::{L2BlockSealStage, L2_BLOCK_METRICS},
+    updates::L2BlockSealCommand,
+};
+
+/// Helper struct that encapsulates parallel l2 block sealing logic.
+#[derive(Debug)]
+pub(crate) struct L2BlockSealProcess;
+
+impl L2BlockSealProcess {
+    pub fn all_subtasks() -> Vec<Box<dyn L2BlockSealSubtask>> {
+        vec![
+            Box::new(MarkTransactionsInL2BlockSubtask),
+            Box::new(InsertStorageLogsSubtask),
+            Box::new(InsertFactoryDepsSubtask),
+            Box::new(InsertTokensSubtask),
+            Box::new(InsertEventsSubtask),
+            Box::new(InsertL2ToL1LogsSubtask),
+        ]
+    }
+
+    pub fn subtasks_len() -> u32 {
+        Self::all_subtasks().len() as u32
+    }
+
+    pub async fn run_subtasks(
+        command: Arc<L2BlockSealCommand>,
+        pool: ConnectionPool<Core>,
+    ) -> anyhow::Result<()> {
+        let subtasks = Self::all_subtasks();
+        let handles = subtasks.into_iter().map(|subtask| {
+            let command = command.clone();
+            let pool = pool.clone();
+            tokio::spawn(async move {
```

Comment:
Spawning tasks looks redundant; you can create non-`'static` futures in the iterator and join them:

```rust
pub async fn run_subtasks(
    command: &L2BlockSealCommand,
    pool: &ConnectionPool<Core>,
) -> anyhow::Result<()> {
    let subtasks = Self::all_subtasks();
    let handles = subtasks.into_iter().map(|subtask| {
        let subtask_name = subtask.name();
        async move { subtask.run(command, pool).await.context(subtask_name) }
    });
    futures::future::try_join_all(handles).await?;
    Ok(())
}
```

(This assumes that `L2BlockSealSubtask` is changed correspondingly to take `command` / `pool` by a reference.)

Reply:
Stupid question: what's the difference between my and your pieces of code?

Reply:
Your version spawns Tokio tasks, which provides indirection and true parallelism (if it's supported by the async runtime, which it is). On the downside, it requires all spawned tasks to be `'static` (note that in my approach joined futures are not `'static`; they reference `command` and `pool`) and complicates error handling. In contrast, my version would run all joined futures in the parent task (of course, they still would be polled concurrently, but if they need real parallelism to work efficiently, this won't be enough).